table.cc

```diff
@@ -1022,15 +1030,6 @@ table::on_compaction_completion(sstables::compaction_completion_desc& desc) {
         tlogger.error("Compacted SSTables deletion failed: {}. Ignored.", std::current_exception());
     }
 
-    // unconditionally remove compacted sstables from _sstables_compacted_but_not_deleted,
-    // or they could stay forever in the set, resulting in deleted files remaining
-    // opened and disk space not being released until shutdown.
-    std::unordered_set<sstables::shared_sstable> s(
-           desc.old_sstables.begin(), desc.old_sstables.end());
-    auto e = boost::range::remove_if(_sstables_compacted_but_not_deleted, [&] (sstables::shared_sstable sst) {
-        return s.contains(sst);
-    });
-    _sstables_compacted_but_not_deleted.erase(e, _sstables_compacted_but_not_deleted.end());
     rebuild_statistics();
```

Comment:
The call here to rebuild_statistics expects `_sstables_compacted_but_not_deleted` to be already updated so it should also go into the deferred action.

By the way, on the gate closed error path, these sstables aren't really deleted.  I'm not sure they are being tracked correctly, but since this happening on the shutdown path it doesn't matter.

Reply:
Good catch, will update.

Probably, we should delete those sstables unconditionally. The time to delete isn't worth the extra special case. But I won't do that in this series.